File: example/xx_three/billboards.cpp
//#include <mud/frame.h>
#include <frame/Api.h>
#include <gfx-pbr/Api.h>

#include <xx_three/xx_three.h>

#include <stl/vector.hpp>

#include <cstring>

void xx_billboards(Shell& app, Widget& parent, Dockbar& dockbar, bool init)
{
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	ui::orbit_controller(viewer);

	Camera& camera = viewer.m_camera;
	camera.m_near = 2.f; camera.m_far = 2000.f;
	camera.m_eye.z = 1000.f;

	Scene& scene = viewer.m_scene;
	//scene.fog = new THREE.FogExp2(0x000000, 0.001);

	static Program& program = app.m_gfx.programs().fetch("point");

	static Texture& sprite = *app.m_gfx.textures().file("sprites/disc.png");

	constexpr size_t num = 10000;
	//constexpr size_t num = 1;

	static Batch* batch = nullptr;
	static Material* material = nullptr;

	struct Instance { vec3 pos; float pad0; vec2 scale; float pad1; float pad2; };
	static vector<Instance> instances = vector<Instance>(num);

	if(init)
	{
		Model& model = *app.m_gfx.models().get("point");
		
		//Material& material = new THREE.PointsMaterial({ size: 35, sizeAttenuation : false, map : sprite, alphaTest : 0.5, transparent : true });
		Material& mat = app.m_gfx.materials().create("points", [&](Material& m) {
			m.m_program = &program;
			m.m_solid.m_colour = hsl(1.f, 0.3f, 0.7f);
			m.m_solid.m_colour = &sprite;
		});
		material = &mat;

		for(size_t i = 0; i < num; i++)
		{
			vec3 pos = vec3(randf(), randf(), randf()) * 2000.f - 1000.f;
			vec2 scale = vec2(1.f);
			instances[i] = { pos, 0.f, scale, 0.f, 0.f };
		}

		Node3& n = gfx::nodes(scene).add(Node3());
		Item& it = gfx::items(scene).add(Item(n, model, 0U, &mat));

		batch = &gfx::batches(scene).add(Batch(it, sizeof(Instance)));
		it.m_batch = batch;
	}

	//ui::slider_field(viewer, "sizeAttenuation", { material.sizeAttenuation, { 0.f, 1000.f, 1.f } })

	float time = app.m_gfx.m_time;

	static vec2 mouse = vec2(0.f);
	if(MouseEvent event = viewer.mouse_event(DeviceType::Mouse, EventType::Moved))
	{
		mouse = (event.m_relative - viewer.m_frame.m_size / 2.f);
	}

	viewer.m_camera.m_eye.x += (mouse.x - viewer.m_camera.m_eye.x) * 0.05f;
	viewer.m_camera.m_eye.y += (-mouse.y - viewer.m_camera.m_eye.y) * 0.05f;

	//viewer.m_camera.m_target = scene.position;

	float h = fmod(360.f * (1.f + time), 360.f) / 360.f;
	material->m_solid.m_colour = hsl(h, 0.5f, 0.5f);

	span<float> memory = batch->begin(instances.size());
	memcpy(memory.data(), instances.data(), memory.size() * sizeof(float));
}

File: example/xx_three/two/billboards.js

var viewer = two.ui.scene_viewer(panel);
two.ui.orbit_controller(viewer);

var camera = viewer.camera;
camera.near = 2.0; camera.far = 2000.0;
camera.eye.z = 1000.0;

var scene = viewer.scene;
//scene.fog = new THREE.FogExp2(0x000000, 0.001);

var num = 10000;
//var num = 1;

this.batch = nullptr;
//struct Instance { vec4 d0; vec4 d1; };

if (init) {
    this.program = app.gfx.programs.fetch('point');

    this.sprite = app.gfx.textures.file('sprites/disc.png');

    this.material = app.gfx.materials.create('points');
    
    var m = this.material;
    m.program = program;
    m.solid.colour.value = two.hsl(1.0, 0.3, 0.7);
    m.solid.colour.texture = sprite;
    m.point.size = 35.0;
    m.point.project = false;
    m.alpha.alpha_test = 0.5;
    m.alpha.is_alpha = true;

    var model = app.gfx.models.get('point');

    this.instances = [];
    
    for(var i = 0; i < num; i++)
    {
        var pos = new two.vec3(Math.random(), Math.random(), Math.random()) * 2000.0 - 1000.0;
        var scale = new two.vec2(1.0);
        this.instances[i] = { d0: new two.vec4(pos, 0.0), d1: new two.vec4(scale, new two.vec2(0.0)) };
    }

    var n = scene.nodes().add(new two.Node3());
    var it = scene.items().add(new two.Item(n, model, 0, material));

    this.batch = two.gfx.batches(scene).add(new two.Batch(it));
    it.batch = this.batch;
}

//two.ui.slider_field(viewer, 'sizeAttenuation', { material.sizeAttenuation, { 0.0, 1000.0, 1.0 } })

var time = app.gfx.time;

this.mouse = new two.vec2(0.0);
if(var event = viewer.mouse_event(two.DeviceType.Mouse, two.EventType.Moved))
{
    mouse = (event.relative - viewer.frame.size / 2.0);
}

viewer.camera.eye.x += (mouse.x - viewer.camera.eye.x) * 0.05;
viewer.camera.eye.y += (-mouse.y - viewer.camera.eye.y) * 0.05;

//viewer.camera.target = scene.position;

var h = ((360.0 * (1.0 + time)) % 360.0) / 360.0;
this.material.solid.colour = two.hsl(h, 0.5, 0.5);

span<float> memory = batch->begin(instances.length, sizeof(Instance));
memcpy(memory.data(), instances.data(), memory.length * sizeof(float));
